### PR TITLE
[scala-stream-collector] Expose metrics over jmx

### PR DESCRIPTION
**Pull request by jspc**
The ability to determine whether or not the stream collector is doing much without interrogating/ pattern matching logs is useful for scaling events and debugging.

This PR introduces a custom mbean which starts to expose useful information. Hooking it into `jmxtrans` with the following config (for example):

```xml
    <query objectName="com.snowplowanalytics.snowplow.collectors.scalastream.monitoring:type=ScalaCollector" attribute="Latest"
           resultAlias="scalastream.Latest" />
    <query objectName="com.snowplowanalytics.snowplow.collectors.scalastream.monitoring:type=ScalaCollector" attribute="Requests"
           resultAlias="scalastream.Requests" />
    <query objectName="com.snowplowanalytics.snowplow.collectors.scalastream.monitoring:type=ScalaCollector" attribute="SuccessfulRequests"
           resultAlias="scalastream.SuccessfulRequests"/>
    <query objectName="com.snowplowanalytics.snowplow.collectors.scalastream.monitoring:type=ScalaCollector" attribute="FailedRequests"
           resultAlias="scalastream.FailedRequests"/>
```

Would allow a developer to interrogate metrics like:

```
scalastream.Latest 1519302379 1519302389
scalastream.Requests 1 1519302389
scalastream.SuccessfulRequests 0 1519302389
scalastream.FailedRequests 1 1519302389
```

### CODE REVIEW

**Comment by jspc** [comment]
Actually- this PR is crap; I'm blindly assuming that all non-404 requests are fine.

Let me fix this.

**Comment by jspc** [comment]
@BenFradet - thanks for assigning yourself to this PR. I've sorted the thing I caught earlier.

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala [hunk 1]
```diff
@@ -43,30 +46,38 @@ trait CollectorRoute {
...
                 entity(as[String]) { body =>
-                  complete {
-                    collectorService.cookie(qs, Some(body),
-                      path, cookie, userAgent, refererURI, host, ip, request, false, Some(ct))._1
-                  }
+                  val (r,l) = collectorService.cookie(qs, Some(body),
+                      path, cookie, userAgent, refererURI, host, ip, request, false, Some(ct))
+
+                  incrementRequests(r._1)
+                  complete {r}
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala]: you can do:

```scala
val (r, _) = collectorService.cookie(qs, Some(body),
  path, cookie, userAgent, refererURI, host, ip, request, false, Some(ct))
incrementRequests(r.status)
complete(r)
```

same thing below

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala [hunk 2]
```diff
@@ -14,14 +14,17 @@
...
+import akka.http.scaladsl.model.{ContentType, HttpResponse, StatusCode, StatusCodes}
 import akka.http.scaladsl.model.headers.HttpCookiePair
 import akka.http.scaladsl.server.{Directive1, Route}
 import akka.http.scaladsl.server.Directives._
+import com.snowplowanalytics.snowplow.collectors.scalastream.monitoring.BeanRegistry
 
 trait CollectorRoute {
   def collectorService: Service
 
+  private val successfulStatuses = List(StatusCodes.OK, StatusCodes.Found)
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala]: can we move this next to `incrementRequests` so we know what it relates to

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala [hunk 3]
```diff
@@ -14,14 +14,17 @@
  */
 package com.snowplowanalytics.snowplow.collectors.scalastream
 
-import akka.http.scaladsl.model.{ContentType, HttpResponse}
+import akka.http.scaladsl.model.{ContentType, HttpResponse, StatusCode, StatusCodes}
 import akka.http.scaladsl.model.headers.HttpCookiePair
 import akka.http.scaladsl.server.{Directive1, Route}
 import akka.http.scaladsl.server.Directives._
+import com.snowplowanalytics.snowplow.collectors.scalastream.monitoring.BeanRegistry
```
Comment (BenFradet) [inline 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala]: you can `import monitoring.BeanRegistry`

also we try to separate internal from external imports by a newline, e.g. https://github.com/jspc/snowplow/blob/30d01e8909ffb76239759bdedb70ee473cecd4d0/2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/Collector.scala#L27-L29

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

📝 Please visit https://github.com/snowplow/snowplow/wiki/CLA to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala (hunk 3 above)
Comment (jspc) [reply 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala]: I was in two minds about the relative import, but I'm happy to follow your lead.

**Comment by jspc** [comment]
I have signed the CLA

**Comment by snowplowcla** [comment]
Confirmed! @jspc has signed the Individual Contributor License Agreement. Thanks so much

**Comment by BenFradet** [comment]
Incorporated into a1051070e6bee23e225a000473808d4fc4178c65, closing

**Pull request closed without merging**